pkg/virtual/generic/rootapiserver/root_apiserver.go

```diff
@@ -0,0 +1,187 @@
+package rootapiserver
+
+import (
+	"context"
+	"fmt"
+	// "net/http"
+	"time"
+
+	// corev1 "k8s.io/api/core/v1"
+	// kapierror "k8s.io/apimachinery/pkg/api/errors"
+	// metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	// utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+
+	// genericmux "k8s.io/apiserver/pkg/server/mux"
+	kubeinformers "k8s.io/client-go/informers"
+	// corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+
+	// rbacrest "k8s.io/kubernetes/pkg/registry/rbac/rest"
+	rbacregistryvalidation "k8s.io/kubernetes/pkg/registry/rbac/validation"
+	rbacauthorizer "k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac"
+
+	virtualapiserver "github.com/kcp-dev/kcp/pkg/virtual/generic/apiserver"
+)
+
+type RootAPIExtraConfig struct {
+	// we phrase it like this so we can build the post-start-hook, but no one can take more indirect dependencies on informers
+	InformerStart func(stopCh <-chan struct{})
+
+	KubeAPIServerClientConfig *rest.Config
+	KubeInformers             kubeinformers.SharedInformerFactory
+
+	GroupAPIServerBuilders []GroupAPIServerBuilder
+	RootPathResolver func(urlPath string, context context.Context) (bool, context.Context)
+
+	// these are all required to build our storage
+	RuleResolver   rbacregistryvalidation.AuthorizationRuleResolver
+	SubjectLocator rbacauthorizer.SubjectLocator
+
+	RESTMapper                *restmapper.DeferredDiscoveryRESTMapper
+}
+
+// Validate helps ensure that we build this config correctly, because there are lots of bits to remember for now
+func (c *RootAPIExtraConfig) Validate() error {
+	ret := []error{}
+
+	if c.KubeInformers == nil {
+		ret = append(ret, fmt.Errorf("KubeInformers is required"))
+	}
+	if c.RuleResolver == nil {
+		ret = append(ret, fmt.Errorf("RuleResolver is required"))
+	}
+	if c.SubjectLocator == nil {
+		ret = append(ret, fmt.Errorf("SubjectLocator is required"))
+	}
+	if c.RESTMapper == nil {
+		ret = append(ret, fmt.Errorf("RESTMapper is required"))
+	}
+
+	return utilerrors.NewAggregate(ret)
+}
+
+type RootAPIConfig struct {
```

Comment:
I'm getting a bit lost in all the 30 `*Config` structs - why all the complexity?